07_classes.md

```diff
@@ -0,0 +1,565 @@
+
+## Chapter objectives
+
+Classes are a very important concept for writing a full-fledged application in Python rather than just a little script because they will help keep your code clean and robust.
+
+In this chapter you will learn:
+
+- What a "class" is and why they are useful
+- What "instance", "attribute" and "method" mean
+- How to define and use classes
+    - Classes can get very complex but we will cover the core features
+
+## What is a class?
+
+In Python 3, a class is a type; a type is a class. The only difference is that people tend to refer to the built-in types as "types" and user-defined types as "classes". There can be more of a distinction in other languages (including Python 2).
+
+> For a recap on "types", refer back to Chapter 3.
+
+Say you define a `Dog` class. This is where you decide what it means to be a Dog
+- What data is associated with each Dog?
+- What can a Dog do?
+
+Someone's pet dog is an **instance** of that class (an object of type `Dog`).
+
+Data or functions can be stored on the class itself or on individual instances. But in general, the class will define how the object looks/behaves and then you will provide specific data to an instance.
+
+There are multiple reasons classes are useful. They will help you to follow some of the principles of "good" object-oriented programming, such as the five "SOLID" principles, which are covered in detail in module 2 of the course. But here are two fundamental uses of classes:
+
+### Group variables together into entities
+
+You don't want to pass around multiple variables (like `dog_name`, `owner`, `breed` etc) for each dog that passes through the system. So group them together into a single entity (a `Dog`). Data classes are simple classes that hold data and don't necessarily provide any functionality beyond that.
+
+This _could_ be achieved with a dictionary but a class is more robust. The class definition is a single source of truth for what it means to be a Dog. You can ensure that every Dog in your system has the same shape. Dictionaries are less reliable - key/value pairs can be added/removed at any point and it's not obvious where to look to find out what your dictionary must/can contain.
+
+You can also for example make some of the dog's data constant or optional, without causing bugs because it's clearly laid out in the definition of Dog.
+
+### Linking functionality with state
+
+In Python, you could put all of your dog-related functions in a "dog_functions.py" module. But putting it into the Dog class could be nicer. E.g. say you want a function `is_healthy_weight` that tells you whether an individual dog is healthy based on the data you hold about it. Putting it in a module means you need something like this:
+
+```python
+import dog_functions
+
+my_dog_is_healthy = dog_functions.is_healthy_weight(my_dog)
+```
+
+If the function is part of the Dog class, then using it looks like this:
+
+```python
+my_dog_is_healthy = my_dog.is_healthy_weight()
+```
```

Comment:
I'd be tempted to put this after the `### Methods` section below so we get to defining classes quicker? Fine either way